### PR TITLE
Demonstration of Extension Approach 2

### PR DESCRIPTION
**Pull request by nimish-mehta**

### CODE REVIEW

File: apps/extensions_manager/lib/extensions_manager/model_extension.ex [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        end
+      end
+
+      def schema_changes do
+        @schema_changes
+      end
+
+      defmacro include_methods do
+        quote do
+          unquote(Enum.map(method_blocks, fn (method_block) -> method_block end))
```
Comment (mgwidmann) [inline apps/extensions_manager/lib/extensions_manager/model_extension.ex:62]: Why not just `unquote(method_blocks)`? The map doesn't look like it makes any effect on the data...

File: apps/extensions_manager/lib/extensions_manager/model_extension.ex [hunk 2]
```diff
@@ -0,0 +1,75 @@
+defmodule ExtensionsManager.ModelExtension do
+  defmacro __using__(_opts) do
+    quote do
+      Module.register_attribute(__MODULE__, :schema_changes, accumulate: true)
+      Module.register_attribute(__MODULE__, :method_block, accumulate: true)
+
+      import ExtensionsManager.ModelExtension, only: [add_to_schema: 4, add_to_schema: 3, include_method: 1]
+      @before_compile ExtensionsManager.ModelExtension
+
+      defmacro __using__(_opts) do
```
Comment (mgwidmann) [inline apps/extensions_manager/lib/extensions_manager/model_extension.ex:10]: Does defining a `__using__/1` prevent the user from defining their own `__using__/1`? Since most `use X` statements are at the top, that'd mean this one would match everything and any other `__using__/1` definitions would generate a compiler warning and never match.

File: apps/favorite_products/lib/favorite_products/nectar_extension.ex [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  end
+
+  defp do_install("products") do
+    quote do
+      ## In Phoenix App Model Schema definition, join association is defined first and then through association
+      ## Please note the reverse order here as while collecting, it gets the reverse order and
+      ## injected into actual model schema as expected .. tricky huhh !!
+      add_to_schema(:has_many, :liked_by, through: [:likes, :user])
+      add_to_schema(:has_many, :likes, FavoriteProducts.UserLike, [])
+      include_method do
```
Comment (mgwidmann) [inline apps/favorite_products/lib/favorite_products/nectar_extension.ex:25]: Maybe I'm not following but why is this necessary? Can't you just do something like

``` elixir
quote do
  def like_changest(...)
end
```

File: apps/favorite_products/lib/favorite_products/nectar_extension.ex [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    end
+  end
+
+  defp do_install("products") do
+    quote do
+      ## In Phoenix App Model Schema definition, join association is defined first and then through association
+      ## Please note the reverse order here as while collecting, it gets the reverse order and
+      ## injected into actual model schema as expected .. tricky huhh !!
+      add_to_schema(:has_many, :liked_by, through: [:likes, :user])
+      add_to_schema(:has_many, :likes, FavoriteProducts.UserLike, [])
```
Comment (mgwidmann) [inline apps/favorite_products/lib/favorite_products/nectar_extension.ex:24]: All this seems to do is call `has_many :likes, FavoriteProducts.UserLike`... Maybe this is too much magic? :crystal_ball:

File: apps/favorite_products/priv/static/js/phoenix.js [hunk 5]
```diff
@@ -0,0 +1,1026 @@
+(function(exports){
```
Comment (mgwidmann) [inline apps/favorite_products/priv/static/js/phoenix.js:1]: Newer versions of Phoenix have this file pulled directly from deps, preventing you from needing to check it in. See the upgrade guide: https://gist.github.com/chrismccord/d5bc5f8e38c8f76cad33

File: apps/extensions_manager/lib/extensions_manager/model_extension.ex (hunk 1 above)
Comment (nimish-mehta) [inline apps/extensions_manager/lib/extensions_manager/model_extension.ex:62]: Thanks for pointing this out, yes it can be safely removed with no impact. :+1:

File: apps/favorite_products/lib/favorite_products/nectar_extension.ex (hunk 4 above)
Comment (nimish-mehta) [inline apps/favorite_products/lib/favorite_products/nectar_extension.ex:24]: Yeah, we should probably reduce the amount of magic here and advocate more `explicitness over implicitness`

What do you think of `extend_schema` (sample below) and, then we can collect all declarations similar to `include_method` blocks and extend the Ecto Schema :bulb:

``` elixir
extend_schema do
  has_many :likes, FavoriteProducts.UserLike, [])
  has_many :liked_by, through: [:likes, :user])
end
```

File: apps/favorite_products/priv/static/js/phoenix.js (hunk 5 above)
Comment (nimish-mehta) [inline apps/favorite_products/priv/static/js/phoenix.js:1]: Thanks for the heads-up, going through the update guide.

File: apps/extensions_manager/lib/extensions_manager/model_extension.ex (hunk 2 above)
Comment (pikender) [inline apps/extensions_manager/lib/extensions_manager/model_extension.ex:10]: Yes, Unfortunately this will prevent the another **using**/1 declaration in `ExtensionsManager.ExtendUser` (where its currently being used)

`ExtensionsManager.ExtendUser` is a special module wrapping up _schema changes_ and _support functions_ for `Nectar.User` (Ecto Model) from other modules (extensions) so we never envisaged a situation where another **using** macro would be required for `ExtensionsManager.ExtendUser`

But for exceptional cases, we can probably add a `defoverridable [__using__: 1]` in `ExtensionsManager.ExtendUser` trusting the developer's intent for change :thought_balloon: :guardsman:

File: apps/favorite_products/lib/favorite_products/nectar_extension.ex (hunk 3 above)
Comment (pikender) [inline apps/favorite_products/lib/favorite_products/nectar_extension.ex:25]: @mgwidmann, we can make it work as you suggested but prefer to keep it for reasons as mentioned below :blush: :+1: 

`include_method` wrapping is chosen to help with readability and intent of inclusion (support functions here from extensions)

As, we are targeting a framework where we can reason about any additional block and **able to communicate/converse with fellow team members**, we introduced a common term to base our discussions on :blush: 

As we will come across more use-cases of extensions, everything still will be a function helping us achieve the needed requirement / feature and the **scoping of methods under `include_method`, `create_event`, `extend_channel` will help us reason about them in a controlled / defined manner.**

File: apps/extensions_manager/lib/extensions_manager/model_extension.ex (hunk 2 above)
Comment (mgwidmann) [inline apps/extensions_manager/lib/extensions_manager/model_extension.ex:10]: I'm thinking maybe you're attempt to implicitly change the schema definition should take a more explicit approach. Create a macro instead that adds the fields you want and just insert that macro where it needs to go.

File: apps/favorite_products/lib/favorite_products/nectar_extension.ex (hunk 4 above)
Comment (mgwidmann) [inline apps/favorite_products/lib/favorite_products/nectar_extension.ex:24]: Maybe its best to just supply a macro that adds the fields you need and just call it where it needs to go... 0 magic then

File: apps/extensions_manager/lib/extensions_manager/model_extension.ex (hunk 2 above)
Comment (pikender) [inline apps/extensions_manager/lib/extensions_manager/model_extension.ex:10]: @mgwidmann Project is in very early stages. It will evolve where `Nectar` and all other extensions will be phoenix projects which can be downloaded as hex packages.

Nectar and extensions can be grown and develop in isolations. Nectar being maintained by one team and extensions probably by many different teams where no one team will own all the extensions and Nectar Project.

So, Nectar should provide a way to extensions to hook back and grow independently.

Changes in extensions should not result / demand a change in Nectar Project but it could work otherwise, wherein extensions can adapt to change as per Nectar, if need some specific functionality which is provided by Nectar later.

Not very sure, whether above answers your question or not :( :disappointed: 

We are in process of releasing the Blog Series which will provide in-depth details of nature of project and how it would be used.

We hope, it should make things more clear :blush:

File: apps/favorite_products/lib/favorite_products/nectar_extension.ex (hunk 4 above)
Comment (pikender) [inline apps/favorite_products/lib/favorite_products/nectar_extension.ex:24]: Answered above. [Here](https://github.com/vinsol/nectarcommerce/pull/47#discussion_r58419505)

File: apps/extensions_manager/lib/extensions_manager/model_extension.ex (hunk 2 above)
Comment (mgwidmann) [inline apps/extensions_manager/lib/extensions_manager/model_extension.ex:10]: Ok, its your choice. Just wanted to provide some input. I'll look for your blog post.

**Comment by pikender** [comment]
Extensions Manager is merged to master
and Demo branch created for reference to integration of extensions https://github.com/vinsol/nectarcommerce/pull/64

Checkout the Blog Series [here](http://vinsol.github.io/nectarcommerce/)
